SAM2_Tracking/utils.py

```diff
@@ -35,6 +36,215 @@ def read_config_yaml(config_path):
+
+    Examples
+    --------
+    >>> configs = {
+    ...     "fps": [32, 64],
+    ...     "out_fps": 3,
+    ...     "video_frame_size": [[640, 480], [1280, 720]]
+    ... }
+    >>> get_trial_config(configs, 1)
+    {'fps': 64, 'out_fps': 3, 'video_frame_size': [1280, 720]}
+    """
+    
+    trial_config = {}
+    for key, value in configs.items():
+        # Handle "video_frame_size" separately
+        if key == "video_frame_size":
+            # Check for list of lists
+            if not lol_check(value):
+                trial_config[key] = value # Assign the entire value if not a list of lists
+            else: 
+                trial_config[key] = value[i] # Assign the i-th list if not a list of lists
+        elif isinstance(value,list):
+            if len(value) > 1: # Lists with more than one value
+                trial_config[key] = value[i] # Assign the i-th value
+            else: # List with a single value
+                trial_config[key] = value[0] # Assign the first (only) value
+        else: # single non-list value
+            trial_config[key] = value
+    return trial_config
+
+def run_segmentation(config_file, device):
+    """
+    Runs SAM2 segmentation and mask propagation for one or more trial configurations.
+
+    This function reads a YAML configuration file and processes it to extract 
+    parameters for each trial. Each configuration parameter in the YAML file must 
+    either be a single value (applied to all trials) or a list of values (one per trial). 
+    For each trial, the function initializes a `SAM2FishSegmenter` with the appropriate 
+    configuration and executes the segmentation and propagation process.
+
+    Parameters
+    ----------
+    config_file : str
+        Path to the YAML configuration file containing all segmentation parameters.
+        Each parameter should either be a scalar (applied to all trials) or a list of values
+        (with one entry per trial).
+    device : str
+        The device on which to run the segmentation model (e.g., "cuda" or "cpu").
+
+    Returns
+    -------
+    None
+        The function does not return anything explicitly. However, it saves a pickled 
+        dictionary of masks (as specified by `masks_dict_file` in each trial config) 
+        for each trial after segmentation and propagation.
+
+    Notes
+    -----
+    - The number of trials is determined by the length of list-type parameters in the 
+      configuration file. All such parameters must have the same length.
+    - The function uses `read_config_yaml`, `extract_config_lens`, and `get_trial_config` 
+      as helpers in utils to parse and manage configurations.
+    
+    Warnings
+    --------
+    If the configuration specifies only a single `masks_dict_file` name while running multiple 
+    trials, the output masks will be written to the same file, and results will be 
+    overwritten. To prevent this, provide a list of unique `masks_dict_file` names — 
```

Comment:
This is another reason why we should implement the checking function we discussed. I can see this happening easily.